[2단계 - 자동차 경주 구현] 조조그린(조한석) 미션 제출합니다.

jojogreen91 opened the pull request:
### 안녕하세요 스티치😀
2단계를 하면서 코드에 많은 변화가 있었어요.
1단계 리뷰에서 마무리 짓지 못한 내용과 2단계 미션인 MVC 패턴 적용
그리고 개인적으로 바꾸고 싶은 부분들을 적용하다 보니 코드 수정이 많아졌네요.

2단계에서 작업한 내용을 남겨 놓겠습니다 :)

**1. MVC 패턴 적용**
- 1단계 미션에서 어느정도 `MVC` 패턴을 참고해서 구현 했기때문에 많이 바꾸지는 않았지만 패키지 정리와 `domain`, `controller`, `view` 의 관계를 한번 더 점검해봤습니다.

**2. 정적 팩토리 메서드 제거**
- 스티치 리뷰에 남겨주신 말대로 `Car` 클래스의 정적팩토리 메서드는 필요없게 느껴져서 제거했습니다 :)

**3. `getAsInt()` -> `orElse()`**
- 스트림 사용에 있어서 `getAsInt()` 대신 `orElse()` 를 사용했습니다. `NPE`이 발생하는것을 막는 방향으로 수정했습니다.
- 고민되는 점은 `orElseThrow()` 를 사용해서 적절한 예외를 발생시키는 것과 그냥 일단 에러없이 프로그램이 돌아가도록 `orElse()` 를 쓰는것 중에 어떤게 적절한 방법인지 헷갈리는것 같아요.

**4. `RacingGame` 생성자 수정**
- 스티치의 피드백을 적용하기 위해서(`RacingGame` 멤버변수 초기화 문제) `RacingGame` 객체가 생성될 때 사용자에게서 값을 입력 받아 `Cars` 와 `TryCount` 를 초기화 해주는 방식으로 수정했습니다.(특히 현재 racingcar 프로그램 상황에서)

**5. 예외상황에서 다시 입력을 받는 방식 변경**
- 4번의 내용을 작업하면서 기존의 입력 방식하고 잘 안 맞았고, 재귀가 아닌 방식으로 예외상황에서의 반복적 입력을 구현 하고 싶어서 코드를 수정했습니다.

**6. 게임 결과 관련 코드 변경**
- 기존의 코드의 게임 결과를 반환하는 방식이 스티치의 피드백 대로 다소 애매한 역할을 가지는 것 같아서 새롭게 만들었습니다.
- `GameLog` 클래스를 만들어서 한 턴이 진행될 때 마다 자동차들의 상태를 기록하도록 했습니다.
- 이제 중간결과, 최종게임결과는 `tryCount` 를 매겨변수로 이용해서 `GameLog` 의 메서드를 통해 반환 받을 수 있습니다.
- 더불어 객체의 복사를 위해서 `Car` 클래스에 `Cloneable` 인터페이스를 상속받아 `.clone()` 를 오버라이딩했습니다. (`Car` 객체의 필드는 모두 `immutable` 하고 `primitive` 한 타입이라 `.clone()` 을 통해 깊은 복사가 가능했습니다.)

**7. 테스트 코드 정비**
- 바뀐 코드에 따른 테스트 코드도 다시한번 정비 했습니다.

떨리는 마음으로 피드백 기다리고 있겠습니다.
감사합니다 스티치 😋

jojogreen91 committed:
refactor(RacingGame): change category

jojogreen91 committed:
refactor: change class name

jojogreen91 committed:
refactor(Controller): change category

jojogreen91 committed:
feat: make Application class

jojogreen91 committed:
refactor: delete method

jojogreen91 committed:
fix: fix method for preventing NPE

jojogreen91 committed:
refactor: change variable name

jojogreen91 committed:
refactor(RacingGame): change category

jojogreen91 committed:
refactor: change structure of RacingGame class
- RacingGame 객체가 생성될 때 자동차이름과 시도 횟수를 입력 받는 방식으로 구조를 변경했다.

jojogreen91 committed:
refactor(Controller): add RacingGame as field

jojogreen91 committed:
refactor: change input retry rogic
- 입력되는 값이 예외를 발생 시킬때 다시 값을 입력 받는 코드를 다른 방식으로 교체했다.

jojogreen91 committed:
refactor(TryCount): change method name

jojogreen91 committed:
fix: remake game result logic
- 게임 결과를 반환과 관련된 로직을 변경했다.
- GameLog 클래스를 만들어서 게임이 진행되는 로그를 저장하도록 했다.

jojogreen91 committed:
refactor(GameLog): change methods
- 메서드 이름과 매개변수를 좀더 가독성 있게 변경했다.
- 필요없는 기능을 통폐합했다.

jojogreen91 committed:
test: add RacingGame class test

jojogreen91 committed:
test: add GameLog class test

jojogreen91 committed:
test(GameLogTest): add test

jojogreen91 committed:
refactor: change method name

lxxjn0 commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,58 +1,12 @@
-
-    private void handleException(Exception e) {
-        view.printErrorMessage(e.getMessage());
+        Controller controller = new Controller(new RacingcarRandomNumberGenerator());
```
웹 애플리케이션의 관점에서 보면 MVC 패턴의 Controller는 외부의 요청과 그에 대한 입력을 받아서 내부 로직으로 처리를 하고, 처리된 결과를 사용자에게 돌려주는 역할을 합니다.

그런 관점에서 한번 Controller로 요청을 보내보고 그 결과를 출력하는 식으로 코드를 개선해보는 것도 좋을 것 같습니다!! 
(간단하게 힌트를 하나 드리자면 제가 말한 형태로 코드를 작성해보려면 우선 입력을 controller로 전달해야 합니다. 그렇다면 입력값을 application에서 받아서 controller로 넘겨주는 구조가 될 수도 있겠죠?)

lxxjn0 replied:
추가로 위에 제가 제안드린 방법의 경우 정답은 아닐수도 있어요!! application 로직을 관리하는 부분을 controller로 볼 수도 있습니다!! 단지 저의 생각으로는 웹 애플리케이션과 동일한 구조로 작성해본다면 위의 제안한 구조도 재밌는 방향이 될 수 있을 것 같다는 생각이 들어서 제안드려봤어요!! :)

jojogreen91 replied:
말씀하신대로 `Application` `main` 안에서 `Controller` 가 입력을 전달 받고 `Controller.runGame()` 을 통해서 결과를 출력받는 방식으로 수정해봤습니다!

lxxjn0 commented on src/main/java/racingcar/controller/Controller.java:
```diff
@@ -0,0 +1,72 @@
+    private final RacingGame racingGame;
+
+    private String receivedNamesForGame = INPUT_INIT_VALUE;
+    private String receivedTryCountForGame = INPUT_INIT_VALUE;
```
해당 값들은 꼭 맴버 변수로 관리되어야 할까요?? Controller가 가져야 할 속성인지 한 번 고려해보면 좋을 것 같아요!!
또한 코드 상으로도 로컬 변수로 충분히 풀어낼 수 있을 것 같아보이네요 :)

lxxjn0 replied:
또한 코드 상에 의미없는 값일 경우에는 오히려 초기화 하지 않고 사용하는 시점에서 null 체크를 하는 것이 더 좋아보입니다!! 저의 경우도 `"init"` 이라는 문자열이 다른 로직 상의 처리를 위해 사용되는 값으로 생각하고 다른 코드들을 많이 돌아봤었어요.

jojogreen91 replied:
수정했습니다🔧

lxxjn0 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,21 +1,18 @@
 package racingcar.domain;
 
-public class Car {
+public class Car implements Cloneable {
```
Car 클래스에 Cloneable 인터페이스를 구현하셨군요!! 지난번에 제가 남긴 코멘트를 보고 이렇게 구현하신 것 같은데 맞을까요??

제가 당시 코멘트에서 Cloneable 인터페이스를 말씀드린 이유는 deep copy의 예시를 위해 말씀드린 것이긴 했어요.. 😭 제가 듣기로 Cloneable 인터페이스는 실제로는 많이 사용되지 않고 있는 인터페이스라고 들었습니다. 아마 clone 메서드 자체가 객체 전체를 복사하다 보니 무거운 로직이여서 그렇다고 알고는 있어요. 물론 이렇게 한번 구현해보는 것도 좋은 경험이라고 생각합니다!!

제가 괜히 2번 일을 시키는 것 같아 죄송하지만.. 이번엔 한번 Cloneable 인터페이스를 사용하지 않고 GameLog에 자동차의 이름과 위치를 저장하는 방법을 고민해보는 것은 어떨까요?? 이렇게 도메인에 Cloneable 인터페이스를 구현하지 않아도 충분히 가능한 방법이 있을 것 같아요!! 

이번에 한번 시도해보시고 혹시 어려우시다면 제가 다음 피드백에서 힌트를 드려보도록 할게요!! 기대해보겠습니다!! 아자아자!!! 👍🏼

jojogreen91 replied:
많이 고민했는데 결국엔 `CarForLog` 라는 기록용 객체를 만들어서 기록용 객체를 `GameLog` 에 기록하는 방식을 선택 했습니다. 좋은 방법인지는 모르겠네요...ㅠㅠ

lxxjn0 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -2,51 +2,37 @@
 public class Cars {
 
-    private final List<Car> cars;
+    private final List<Car> carList;
```
혹시 해당 맴버 변수의 이름을 변경한 이유가 따로 있을까요?? 아래의 글을 읽어보시면 좋을 것 같아요!!

> [좋은 코드를 위한 자바 메서드 네이밍](https://tecoble.techcourse.co.kr/post/2020-04-26-Method-Naming/)

jojogreen91 replied:
처음엔 `cars` 라고 한것을 `carList` 가 개인적으로 더 이해하기 편한 것 같아서 한번 바꿔봤는데 변수이름에 `List` 같은 자료형이 들어가는건 별로 좋지 않은 방식이군요...생각해보니 `Cars` 는 일급 컬렉션 이니까 `cars` 가 더 맞는것 같네요! 수정했습니다🔧

lxxjn0 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -2,51 +2,37 @@
-                .map(Car::of)
+        carList = names.stream()
+                .map(name -> new Car(name))
                 .collect(toList());
```
해당 코드도 메서드 참조에서 이렇게 바꾼 이유가 있을까요???

jojogreen91 replied:
정적 팩토리 메서드를 제거하면서 수정한 코드입니다!

lxxjn0 replied:
해당 코드도 `.map(Car::new)` 와 같이 메서드 참조를 사용할 수 있지 않을까요?

jojogreen91 replied:
수정했습니다🔧

lxxjn0 commented on src/main/java/racingcar/domain/TryCount.java:
```diff
@@ -2,18 +2,26 @@
-    private final int maxTryCount;
-    private int currentTryCount;
+    private final int totalTryCount;
+    private int currentTryCount = 0;
```
해당 코드는 Car 클래스에서와는 다르게 초기화를 하고 있네요!! Car 클래스처럼 생성자에서 초기화를 해주는 것도 좋을 것 같아요!!

jojogreen91 replied:
수정했습니다🔧

lxxjn0 commented on src/main/java/racingcar/domain/game/GameLog.java:
```diff
@@ -0,0 +1,53 @@
+
+public class GameLog {
+
+    private final HashMap<Integer, List<Car>> log;
```
해당 변수를 HashMap 타입으로 선언한 이유가 있나요? Map 타입으로 선언한 것과 HashMap 타입으로 선언한 코드의 차이는 무엇일까요? 한번 공부해보면 좋을 것 같아요!! 👍🏼

jojogreen91 replied:
`interface` 인 `Map` 으로 자료형을 선언하면 해당 변수는 `Map` 을 구현한 여러 타입의 클래스들도 받을 수 있기 때문에 `HashMap` 으로 선언 했을 때보다 유지보수성이 좋다! 가 맞을까요?
변경 했습니다🔧

lxxjn0 replied:
넵넵!! 맞습니다 :) 자바에서는 다양한 인터페이스가 있고, 그것을 구현한 구현체들이 있어요!! 

예를 들면 List 인터페이스를 구현한 LinkedList 구현체와 ArrayList 구현체 등이 있죠!! 해당 구현체들은 List가 제공해야 할 기능들을 제공하면서 동시에 내부적으로는 구현체에서 구현한 자료구조에 따라서 데이터과 관리되고 있습니다! 이러한 코드들을 인터페이스를 활용하여 사용한다면 더 유연하게 코드를 작성하고 관리할 수 있어요!!

자바를 사용하다보면 이처럼 인터페이스를 자연스럽게 잘 활용하는 것은 매우 중요합니다!! 이런 부분에 대해서도 공부를 잘 해보면 많은 도움이 될 것 같아요!! 👍🏼

jojogreen91 replied:
넵! 감사합니다😋

lxxjn0 commented on src/main/java/racingcar/domain/receiver/CarNamesReceiver.java:
```diff
@@ -1,30 +1,29 @@
 import racingcar.exception.CarNameException;
 
-public class CarNameParser {
+public class CarNamesReceiver {
```
해당 코드는 도메인 코드라고 생각해서 domain 패키지 하위로 분리하신 걸까요? 

사실 어떠한 로직이 도메인에 해당하는지 여부는 관점에 따라 다르게 볼 수 있다고 생각합니다. 지금처럼 이름을 정하는 규칙 또한 게임의 규칙이라고 생각하고 관리할 수 있고요. 이번 코멘트는 단지 제가 생각하는 도메인 로직은 어떤 것인지에 대한 의견이라고 생각해주시면 좋을 것 같습니다!!

해당 코드에서만 기준으로 본다면 이름이 공백이 있는지, 5자가 넘는지, 이름에 중복이 존재하는지 여부가 관리되고 있는 코드로 보여요. 만약 '이름' 이라는 도메인으로 관리하고 싶다면, 저는 '이름이라는 도메인은 공백이면 안되고 5자가 넘어가면 안된다' 정도로 생각해볼 수 있을 것 같아요. 

그렇다면 이름끼리 중복이 발생하는 것은 '이름' 도메인의 문제일까요? 아니면 입력 값의 문제일까요? 그렇다면 중복된 이름에 대한 유효성은 입력 시에 처리를 했어야 할까요? 입력을 받았을 때 이름을 `,`에 맞게 split 하는 로직도 과연 도메인이 관리해야 하는 로직일까요?

사실 이 경계는 너무 모호하다고 생각하는데 저의 경우는 이름을 입력받는 지점에서 중복 여부를 처리하는 것은 입력 유효성의 검증, 이름을 생성할 때 공백이거나 5자 이상인지 여부를 판단하는 것은 도메인 유효성의 검증이라고 생각이 됩니다. 단순히 문자열을 자르는 로직이 정말 도메인 로직일까? 생각해보면 이 또한 아니라는 생각이 들어요.

해당 코드를 한 번 더 역할에 맞게 분리해보고, 나눠보고, 도메인 객체로도 만들어보고 하면서 조조그린이 생각하는 도메인에 대해 한 번 생각해보면 좋을 것 같습니다 :)

jojogreen91 replied:
스티치 말대로 다시한번 도메인에 대해서 생각해 봤습니다. 제가 생각해도 기존의 `CarNamesReceiver` 는 뭔가 애매한 역할이었다고 생각이 드네요. 그래서 한번 수정해봤습니다!

결과적으로 `CarNamesReceiver` 를 두개의 클래스로 분리 시켜줬습니다.
`NamesInput` 과 `NamesSettingValidator` 로 분리 시켰는데요.

`NamesInput` 은 카테고리를 `view` 쪽으로 옮겼습니다.
게임에 사용될 이름들을 안내 메시지들과 함께, 규칙에 맞을 때 까지 반복적으로 받아주는 역할을 하기 때문에 `view` 쪽이 더 맞다고 생각이 들었습니다.

`NameSettingValidator` 는 `domain` 이 아닌 `validator` 패키지로 묶었습니다.
스티치가 말씀해주신 입력-유효성 검증과 도메인-유효성-검증이 섞여 있는 형태지만 해당 조건들은 애초에 게임을 세팅 하는데 있어서 필수적인 조건(충족하지 않으면 반복해서 재입력을 받아야 되니까) 이라고 생각했습니다.
그래서 `NamesInput` 이 사용자에게 값을 입력받을 때 게임이 세팅 될 수 있는지 검증하는데 사용하도록 했습니다.

lxxjn0 commented on src/test/java/racingcar/domain/game/GameLogTest.java:
```diff
@@ -0,0 +1,48 @@
+    private static final String CAR_1_NAME = "pobi";
+    private static final String CAR_2_NAME = "jason";
+
+    private RacingGame racingGame;
```
현재 GameLog 클래스가 RacingGame에 의해서 진행되고 있습니다. 만약 RacingGame 코드가 문제가 발생한다면 해당 코드들이 모두 테스트가 실패하지 않을까요? 사실 GameLog 코드는 변경된 것도 없고 동작도 정상적으로 동작하는데 말이죠!!

테스트는 최대한 독립되게 실행 되어야 합니다. 특히 단위 테스트의 경우는 더더욱 그렇게 되어야 하고요. GameLog 객체의 메서드에 대한 테스트가 되도록 코드를 수정해보면 좋을 것 같아요!!

jojogreen91 replied:
제가 단위테스트에 대한 개념이 부족했던것 같습니다ㅠ
다시 한번 테스트 코드 작성해봤습니다!

lxxjn0 commented on src/test/java/racingcar/domain/game/RacingGameTest.java:
```diff
@@ -0,0 +1,41 @@
+        racingGame.proceedTurn();
+        racingGame.proceedTurn();
+        assertThat(racingGame.isFinished()).isTrue();
+    }
```
해당 코드도 TryCount 객체에 대한 테스트인가요 RacingGame 객체에 대한 테스트일까요? 사실 TryCount의 increase 함수를 2번 호출한 결과에 대한 테스트나 다름없지 않나요??

단위 테스트는 해당 메서드가 하는 역할에 대한 검증입니다. 해당 테스트 코드는 TryCount 단위 테스트를 작성한다면 이미 검증이 가능한 코드라고 생각합니다. 

단위 테스트가 무엇인지, 어떻게 하면 단위 테스트를 잘 짤 수 있는 구조가 되는지에 대해서 한번 공부해보면 좋을 것 같아요!!

jojogreen91 replied:
수정했습니다🔧

lxxjn0 reviewed (changes requested):
조조그린, 2단계 미션 코드 잘 봤습니다!! :)

1단계 때도 MVC 패턴에 대한 구현은 어느정도 틀이 잡혀있으신 것 같다고 느꼈어요!! 

그래서 이번에는 조금 더 도메인 분리에 대한 고민 포인트에 대한 코멘트와 아직 단위 테스트에 대한 이해가 조금 부족하신 것 같아서 테스트 코드에 대한 피드백을 드렸습니다!! 한번 확인해보시고 반영해주시면 좋을 것 같아요 👍🏼 

그리고 질문하신 부분에 대한 답변입니다!!

3. `orElse()`와 `orElseThrow()`의 사용은 조조그린의 의도에 따라서 사용하면 될 것 같아요!! 만약 해당 로직이 null이 절대 발생하면 안되는 로직이라면 적절한 예외를 발생시켜 프로그램에서 처리할 수 있도록 해야겠죠? 만약 충분히 null이 발생할 수 있는 상황이고, 그런 경우 다르게 처리하면 되는 경우는 예외 대신에 특정한 값을 반환하면 되구요! 해당 코드는 개발자의 의도에 따라서 작성해주시면 될 것 같아요!!!

jojogreen91 committed:
refactor: pull out validator from receiver
- receiver 클래스에서 validator 역할을 분리시켜 꺼냈다.
- Application 에서 Controller 가 생성 될 때 사용자로 부터 입력 받는 구조로 변경하기 위해 receiver 의 역할을 강화 했다.
- receiver 와 validator 를 domain 에서 분리 했다.
- 관련 테스트를 수정했다.

jojogreen91 committed:
refactor(Cars): change field name

jojogreen91 committed:
refactor(TryCount): change Constructor

jojogreen91 committed:
fix(GameLog): change field type

jojogreen91 committed:
refactor: change game log logic
- 게임 로그와 관련된 로직을 수정했다.
- LogCar 라는 게임 기록용 객체를 만들었다.

jojogreen91 committed:
refactor: change class name and category

jojogreen91 committed:
chore: delete line

jojogreen91 committed:
refactor: make View singleton

jojogreen91 committed:
refactor: change class name and catrgory

jojogreen91 committed:
feat: add validate option

jojogreen91 committed:
refactor: change class name

jojogreen91 committed:
refactor: change category

jojogreen91 committed:
test: add Car, Cars test

jojogreen91 committed:
test: add TryCount test

jojogreen91 committed:
test: fix GameLongTest

jojogreen91 committed:
test: fix GameTest

jojogreen91 committed:
test: add @BeforeEach

jojogreen91 commented:
두번째 리뷰 요청 보냅니다 :)
스티치가 말씀해 주신 도메인 분리에 대한 고민과 단위 테스트 구현을 해봤습니다!
나름대로 해본다고 해봤는데 아직 더 연습이 필요한것 같습니다 ㅠㅠ
감사합니다 잘 부탁 드려요 스티치 :)

lxxjn0 commented on src/main/java/racingcar/controller/Controller.java:
```diff
@@ -0,0 +1,26 @@
+            view.printMidtermResults(game.getGameLog(), game.getTryCount().getCurrentTryCount());
+        }
+        view.printWinnersResult(game.getGameLog(), game.getTryCount().getTotalTryCount());
+    }
```
컨트롤러에서는 전체적인 view의 사용과 도메인의 사용에 대한 책임으로 이뤄져 있네요!! 코드가 이전보다 훨씬 적절한 책임으로 나눠져있는 것 같아요!!! 👍🏼

jojogreen91 replied:
감사합니다😋

lxxjn0 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -4,16 +4,14 @@
     public Cars(List<String> names) {
         cars = names.stream()
-                .map(Car::of)
+                .map(name -> new Car(name))
```
해당 부분은 앞선 코멘트와 같이 `.map(Car::new)`와 같이 메서드 참조로 개선이 가능할 것 같아요!!

jojogreen91 replied:
수정했습니다🔧

lxxjn0 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -24,29 +22,12 @@ public void move(RandomNumberGenerator randomNumberGenerator) {
-        return cars.stream()
-                .mapToInt(Car::getPosition)
-                .max()
-                .getAsInt();
```
코드가 많이 분리되면서 Cars가 간결해졌네요!! 👍🏼

jojogreen91 replied:
감사합니다😋

lxxjn0 commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -24,29 +22,12 @@ public void move(RandomNumberGenerator randomNumberGenerator) {
-                .getAsInt();
+    public List<Car> getCars() {
+        return cars;
     }
```
getter가 List<Car>인 cars를 그대로 반환한다면 어떻게 될까요??

```java
List<Car> fieldCars = cars.getCars();
fieldCars.add(new Car(...));
```

외부에서 이렇게 접근해서 Cars 내부 필드로 새로운 Car를 주입할 수 있지 않을까요? 적어도 이런 문제에 대해서 방지하는 방법은 어떤 방법이 있을까요?? 한 번 공부해보고 적용해보면 보다 안전한 코드가 될 수 있을 것 같아요!! 👍🏼

jojogreen91 replied:
`getter` 를 사용하여 일급 컬렉션의 멤버변수인 `cars` 에 직접 접근하던 방식에서
`generateCarForLogList()` 를 이용해 `Cars` 가 직접 `CarForLog` 의 `List` 를 생성하여 응답하는 방식으로 변경했습니다!
`cars` 에 대한 안정성과 상태와 행위를 한곳에서 처리 할 수 있다는 일급 컬렉션의 장점을 더 살릴 수 있는 방법인것 같습니다.

lxxjn0 replied:
아니면 `Collections.unmodifiableList` 메서드도 한번 참고해보면 좋을 것 같아요!!!

lxxjn0 commented on src/main/java/racingcar/domain/Game.java:
```diff
@@ -0,0 +1,42 @@
+import racingcar.domain.random.RandomNumberGenerator;
+import racingcar.exception.RacingGameIsFinishedException;
+
+public class Game {
```
게임에서 게임 진행과 관련된 로직들이 관리되고 있는 것으로 보이네요!! 객체가 적절히 분리되어 있는 점 좋아요 👍🏼

jojogreen91 replied:
감사합니다😋

lxxjn0 commented on src/main/java/racingcar/domain/TryCount.java:
```diff
@@ -2,18 +2,27 @@
     }
 
-    public void increment() {
+    public void increaseCount() {
```
해당 메서드 명도 적절한 동사로 개선하였군요!! 꼼꼼 💯

jojogreen91 replied:
감사합니다😋

lxxjn0 commented on src/main/java/racingcar/domain/log/GameLog.java:
```diff
@@ -0,0 +1,62 @@
+            carForLogList.add(new CarForLog(car.getName(), car.getPosition()));
+        }
+        return carForLogList;
+    }
```
헤딩 코드는 stream을 적절히 사용하면 불변을 활용하여 안전하게 데이터를 생성할 수 있을 것 같아요!!

jojogreen91 replied:
해당 코드는 `Cars` 의 `generateCarForLogList()` 로 옮겨지고 `stream` 으로 작성했습니다!

lxxjn0 commented on src/main/java/racingcar/domain/log/CarForLog.java:
```diff
@@ -0,0 +1,24 @@
+package racingcar.domain.log;
+
+public class CarForLog {
```
제가 의도한 방향에 맞는 구현으로 보입니다!! 👍🏼 

기존에 Car 객체를 사용하였다면 자바의 특성 상 객체 참조에 대한 값 변경의 위험이 항상 존재해요!! 하지만 primitive 타입을 사용하면 그러한 걱정을 하지 않아도 되겠죠? 찰떡으로 알아듣고 멋있게 구현해주었네요!!! 👍🏼

jojogreen91 replied:
많이 고민한 부분인데 다행이네요!ㅠㅠ 감사합니다😋

lxxjn0 commented on src/main/java/racingcar/domain/log/GameLog.java:
```diff
@@ -0,0 +1,62 @@
+import racingcar.domain.Car;
+import racingcar.domain.Cars;
+
+public class GameLog {
```
해당 데이터는 자동차 게임에 대한 도메인 로직이 관리되어 있는 객체인가요?? Log라는 이름에서 알 수 있듯, 기록을 관리하기 위한 데이터 객체로 느껴져요!! 해당 코드는 domain 패키지로 유지되는 것보다는 출력을 관리하는 곳에서 유지되는 것이 더 좋을 것 같아요!! 

그렇다면 View에 출력할 양식에 대한 코드도 해당 객체 내부에 위치할 수 있지 않을까요??

jojogreen91 replied:
`GameLog` 를 `view` 패키지로 옮기고 기존에 `View` 에서 하던 중간결과, 최종결과를 출력하는 책임을 `GameLog` 가 수행하도록 변경해봤어요!
스티치 말대로 `GameLog` 객체는 지금의 역할이 이전보다 더 적절한것 같네요😋

lxxjn0 commented on src/main/java/racingcar/view/View.java:
```diff
@@ -17,26 +17,32 @@
 
-    private Scanner scanner = new Scanner(System.in);
+    private static final View view = new View();
 
-    public String inputCarNames() {
+    private View() {
+    }
+
+    public static View getInstance() {
+        return view;
+    }
```
해당 부분은 View 객체를 하나만 관리하면서 유지하고 싶은 의도로 보여요!! 그렇다면 여기서 싱글톤 패턴을 한 번 공부해보고 적용해보는 것은 어떨까요??

jojogreen91 replied:
현재 싱글톤 패턴으로 적용했다고 생각했는데 혹시 아닌가요?!

lxxjn0 replied:
해당 코드도 크게 차이는 없을 것 같긴 합니다!! 보통 싱글톤 패턴이 인스턴스를 호출했을 때, 없다면 해당 시점에 생성하고 이후부턴 생성한 인스턴스를 반환하는 양식이죠. 그렇다보니 해당 코드랑 조금 스타일이 달라서 말씀드렸습니다 :)

lxxjn0 commented on src/main/java/racingcar/view/input/TryCountInput.java:
```diff
@@ -0,0 +1,43 @@
+
+    private int parseTryCount(String input) {
+        return Integer.parseInt(input);
+    }
```
해당 코드는 새롭게 작성한 메서드 명과 기존 코드의 의미가 동일한 것으로 보여요!! 해당 코드는 따로 메서드 분리를 하지 않아도 될 것 같은데 조조그린 생각은 어떤가요?

jojogreen91 replied:
동의합니다! 수정했습니다🔧

lxxjn0 commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,44 @@
+        cars.move(mockRandomNumberGenerator);
+        cars.move(mockRandomNumberGenerator);
+        cars.move(mockRandomNumberGenerator);
+    }
```
테스트 메서드가 하나 밖에 없어서 굳이 setUp 메서드로 관리하지 않고 메서드 내부에서 처리해줘도 될 것 같아요!!

jojogreen91 replied:
수정했습니다🔧

lxxjn0 commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,44 @@
+                .findFirst()
+                .orElseThrow()
+                .getPosition())
+                .isEqualTo(1);
```
두 검증 로직을 `assertAll()` 메서드로 묶어서 검증하면 좋을 것 같아요!! 이렇게 처리하는 것과 기존에 조조그린이 처리하는 코드의 차이에 대해서도 한번 알아보면 좋을 것 같아요 :)

jojogreen91 replied:
`assertAll()` 로 두가지 검증을 묶어 봤습니다!
여러가지 검증을 하나의 테스트 메서드에서 수행 할 때 하나의 검증이 실패하면 그 상태에서 메서드의 실행이 멈추는데, `asserAll()` 을 사용하면 하나의 검증이 실패 하더라도 멈추지 않고 다른 검증 코드들을 실행한다! 가 맞을까요?

lxxjn0 replied:
넵넵 👍🏼

lxxjn0 reviewed (changes requested):
조조그린, 리뷰 반영한 부분들 잘 봤습니다!! 이제 더는 고칠만한 부분이 거의 없는 것 같아요!! 놀라운 발전 👍🏼 
마지막으로 한 번 고민해보면 좋을 부분들에 대해서 코멘트 남겼어요!! 한번 개선해보고 리뷰 주세요 :)

jojogreen91 committed:
refactor: change to method reference

jojogreen91 committed:
refactor: remove getCars method
- getCars 로 직접적으로 cars 에 접근하는것을 막고 cars 에서 CarForLog 의List 를 생성해서 반환하도록 했다.

jojogreen91 committed:
test: fix CarsTest

jojogreen91 committed:
refactor: delete method and change method name

jojogreen91 committed:
refactor: GameLog change domain to view

jojogreen91 committed:
refactor: replace for logic to stream logic

jojogreen91 committed:
chore: delete lines

jojogreen91 commented:
남겨주신 피드백 반영해봤습니다 :)
프로그램의 도메인 영역에 대한 개념은 아직 헷갈리네요.
그래도 처음 미션을 시작할 때보다는 어렴풋이 보이는것 같기도 하고....ㅋㅋㅋ
스티치 덕분에 많이 배울 수 있었던 것 같습니다.
감사합니다 스티치😀

lxxjn0 reviewed (approved):
조조그린, 자동차 게임 미션 진행하느라 수고하셨습니다 :)

처음 미션 제출했을 때도 정말 1단계 미션이 맞나? 싶은 느낌이 드는 코드였는데 미션이 마무리 될 시점에는 또 그보다 더 발전한 모습을 보여주네요!! 코드에 조조그린의 섬세한 고민들이 느껴져서 인상적이었습니다!! 앞으로도 더 많은 기대 해볼게요 👍🏼 

이번 미션은 여기서 머지하도록 하겠습니다!! 고생하셨어요 :)
